doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -659,168 +734,476 @@ insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.
+
+   :n:`@ltac2_expr`
+      The tactic to apply if the construct matches.  Metavariable values from the pattern
+      match are substituted
+      into :n:`@ltac2_expr` before it's applied.  Note that metavariables are not
+      prefixed with the question mark as they are in :token:`cpattern`.
+
+      If :token:`ltac2_expr` evaluates to a tactic, then it is
+      applied. If the tactic succeeds, the result of the match expression is
+      :n:`()`.  If :token:`ltac2_expr` does not evaluate
+      to a tactic, that value is the result of the match expression.
+
+      If :n:`@ltac2_expr` is a tactic with backtracking points, then subsequent
+      failures after a :tacn:`lazy_match!` or :tacn:`multi_match!` (but not :tacn:`match!`) can cause
+      backtracking into :n:`@ltac2_expr` to select its next success.
+
+   .. note::
+      Each Ltac2 construct is processed in two phases: an evaluation phase and an execution phase.
+      In most cases, tactics that may change the proof state are applied in the second phase.
+      (Tactics that generate integer, string or syntactic values, such as :tacn:`fresh`,
+      are processed during the evaluation phase.)
+
+      Unlike other tactics, `*match*!` tactics get their first success (applying tactics to do
+      so) as part of the evaluation phase.  Among other things, this can affect how early
+      failures are processed in :tacn:`assert_fails`.  Please see the note in :tacn:`assert_fails`.
+
+Variables from the :n:`@tac2pat1` are statically bound in the body of the branch, to
 values of type `constr` for the variables from the :n:`@term` pattern and to a
 value of type `Pattern.context` for the variable :n:`@lident`.
 
 Note that unlike Ltac, only lowercase identifiers are valid as Ltac2
 bindings, so that there will be a syntax error if one of the bound variables
 starts with an uppercase character.
 
-The semantics of this construction is otherwise the same as the corresponding
+The semantics of this construction are otherwise the same as the corresponding
 one from Ltac1, except that it requires the goal to be focused.
 
+.. _ltac2_match_vs_lazymatch_ex:
+
+.. example:: Ltac2 Comparison of lazy_match! and match!
+
+   (Equivalent to this :ref:`Ltac1 example<match_vs_lazymatch_ex>`.)
+
+   These lines define a `msg` tactic that's used in several examples as a more-succinct
+   alternative to `print (to_string "...")`:
+
+   .. From Ltac2 Require Import Message. *outside* the proof lest it disappear with "abort"
+
+   .. coqtop:: in
+
+      From Ltac2 Require Import Message.
+      Ltac2 msg x := print (of_string x).
+
+   .. coqtop:: none
+
+      Goal True.
+
+   In :tacn:`lazy_match!`, if :token:`ltac2_expr` fails, the :tacn:`lazy_match!` fails;
+   it doesn't look for further matches.  In :tacn:`match!`, if :token:`ltac2_expr` fails
+   in a matching branch, it will try to match on subsequent branches.
+
+   .. coqtop:: all
+
+      Fail lazy_match! constr:(True) with
```

Comment:
Do you want to mention somewhere that `'True` is a shorter form of `constr:(True)`?

Reply:
I added a few words.  This should be explained with the syntax for `ltac2_expr` is explained--currently none of that syntax is explained and won't be added in this PR.